File: AndroidClipSync/app/src/main/java/com/mtechdwork/clipsync/Communication.java
package com.mtechdwork.clipsync;

import android.content.Context;
import android.net.DhcpInfo;
import android.net.wifi.WifiManager;
import android.os.StrictMode;
import android.util.Log;

import java.io.IOException;
import java.io.OutputStream;
import java.io.PrintWriter;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.Socket;
import java.nio.charset.StandardCharsets;

public class Communication {

    private final int TCP_PORT = 7071;
    private final Context context;
    private final SettingManager settingManager;
    private DatagramSocket socket = null;

    Communication(Context context) {
        this.context = context.getApplicationContext();
        settingManager = new SettingManager(context);
    }

    private void log(String message, int type) {
        // Type: 0 - Info, 1 - Warning, 2 - Error
        boolean debug = false;
        if (!debug) return;
        String className = "[Communication]";
        switch (type) {
            case 0:
                Log.i(className, message);
                break;

            case 1:
                Log.w(className, message);
                break;

            case 2:
                Log.e(className, message);
        }
    }

    public void sendBroadcast() {
        final String messageStr = "CS_BC_" + settingManager.getUsername(); // ClipSync_Broadcast Flag
        StrictMode.ThreadPolicy policy = new StrictMode.ThreadPolicy.Builder().permitAll().build();
        StrictMode.setThreadPolicy(policy);

        try {
            socket = new DatagramSocket();
            socket.setBroadcast(true);
            byte[] sendData = messageStr.getBytes();
            DatagramPacket sendPacket = new DatagramPacket(sendData, sendData.length, getBroadcastAddress(), 7070);

            for (int i = 0; i < 3; i++) { // Send broadcast 3 times
                socket.send(sendPacket);
                log("Broadcast packet " + (i + 1) + " sent to: " + getBroadcastAddress().getHostAddress(), 0);
                Thread.sleep(100);
            }
        } catch (Exception e) {
            log(e.getMessage(), 2);
        }
        if (socket != null && !socket.isClosed()) socket.close();
    }

    public void sendSyncRequest(InetAddress ipAddress) {
        Authenticator authenticator = new Authenticator(context);
        new Thread(() -> {
            try (Socket socket = new Socket(ipAddress, TCP_PORT);
                 OutputStream outputStream = socket.getOutputStream();
                 PrintWriter writer = new PrintWriter(outputStream, true)) {

                String otp = JsonGen.otpAuthJson(authenticator.genOTP());

                if (otp != null && !otp.isEmpty()) {
                    // Encrypt with XChaCha20-Poly1305
                    byte[] associatedData = "metadata".getBytes(StandardCharsets.UTF_8);
                    String encryptedOTP = XChaChaCrypto.encrypt(otp, associatedData);

                    // Gửi dữ liệu
                    writer.println(encryptedOTP);
                    log("Sent OTP: " + otp, 0);
                    log("Sent encrypted OTP: " + encryptedOTP, 0);
                } else {
                    writer.println("");
                    log("Sent sync request", 0);
                }
            } catch (Exception e) {
                log(e.getMessage(), 2);
            }
        }).start();
    }

    public void sendClipboardData(InetAddress ipAddress) {
        new Thread(() -> {
            try (Socket socket = new Socket(ipAddress, TCP_PORT);
                 OutputStream outputStream = socket.getOutputStream();
                 PrintWriter writer = new PrintWriter(outputStream, true)) {

                String clipboardData = JsonGen.dataJson(ClipboardData.getLatestData());

                if (clipboardData != null) {
                    // Encrypt with XChaCha20-Poly1305
                    byte[] associatedData = "metadata".getBytes(StandardCharsets.UTF_8);
                    String encryptData = XChaChaCrypto.encrypt(clipboardData, associatedData);

                    writer.println(encryptData);
                    log("Sent data: " + clipboardData, 0);
                    log("Sent encrypted data: " + encryptData, 0);
                } else log("Clipboard null. Ignore", 0);
            } catch (Exception e) {
                log(e.getMessage(), 2);
            }
        }).start();
    }

    private InetAddress getBroadcastAddress() throws IOException {
        WifiManager wifi = (WifiManager) context.getApplicationContext().getSystemService(Context.WIFI_SERVICE);
        DhcpInfo dhcp = wifi.getDhcpInfo();

        int broadcast = (dhcp.ipAddress & dhcp.netmask) | ~dhcp.netmask;
        // (&: AND; |: OR; ~: NOT)
        // Example: IP = 192.168.1.10; Netmask = 255.255.255.0
        // IP in base 2: 11000000.10101000.00000001.00001010
        // Netmask in base 2: 11111111.11111111.11111111.00000000
        // => Broadcast IP = (11000000.10101000.00000001.00001010 & 11111111.11111111.11111111.00000000) | ~(11111111.11111111.11111111.00000000)
        // => Broadcast IP = 192.168.1.255

        byte[] quads = new byte[4];
        for (int k = 0; k < 4; k++)
            quads[k] = (byte) ((broadcast >> k * 8) & 0xFF);
        return InetAddress.getByAddress(quads);
    }
}

File: WindowsClipSync/ClipSync/Classes/Communication.vb
Imports System.IO
Imports System.Net
Imports System.Net.Sockets
Imports System.Security.Cryptography.Xml
Imports System.Text
Imports System.Threading

Public Class Communication
	Private Const UDP_PORT = 7070
	Private Const TCP_PORT = 7071

	Public Shared Sub sendBroadcast()
		Dim ipAddr As IPAddress = LocalIP.getIPAddress()

		Dim udpClient As New UdpClient(New IPEndPoint(ipAddr, 0))
		udpClient.EnableBroadcast = True

		Dim data = Encoding.UTF8.GetBytes("CS_BC_" + SettingManager.getUsername())
		Dim endpoint As New IPEndPoint(IPAddress.Broadcast, UDP_PORT)

		For i = 0 To 2 Step 1
			udpClient.Send(data, data.Length, endpoint)
			ClipSyncDebug.log("Broadcast packet " + (i + 1).ToString + " sent to: " + endpoint.Address.ToString)
			Thread.Sleep(100)
		Next
		udpClient.Close()
	End Sub

	Private Shared Sub sendRequestThread(ipAddr As IPAddress)
		Try
			Using socket As New Socket(SocketType.Stream, ProtocolType.Tcp)
				socket.Connect(ipAddr, TCP_PORT)

				Dim otp As String = JsonGen.otpAuthJson(Authenticator.genOTP())

				If otp IsNot Nothing AndAlso otp <> "" Then
					Dim encryptedOTP As String = XChaChaCrypto.encrypt(otp, Encoding.UTF8.GetBytes("metadata"))

					' Send the encrypted OTP
					Using stream As NetworkStream = New NetworkStream(socket)
						Using writer As StreamWriter = New StreamWriter(stream)
							writer.WriteLine(encryptedOTP)
							writer.Flush()
						End Using
					End Using
					ClipSyncDebug.log("[Communication] Sent OTP: " & otp)
					ClipSyncDebug.log("[Communication] Encrypted OTP: " & encryptedOTP)
				Else
					socket.Send(Encoding.UTF8.GetBytes(""))
					ClipSyncDebug.log("[Communication] Send sync request")
				End If
			End Using
		Catch ex As Exception
			ClipSyncDebug.log("[Communication] Sync request failed: " & ex.Message)
		End Try
	End Sub

	Private Shared Sub sendClipboardThread(ipAddr As IPAddress)
		Try
			If ipAddr Is Nothing Then Return
			Using socket As New Socket(SocketType.Stream, ProtocolType.Tcp)
				socket.Connect(ipAddr, TCP_PORT)

				Dim data As String = JsonGen.dataJson(ClipboardData.getLatestData())

				If data IsNot Nothing Then
					Dim encryptedData = XChaChaCrypto.encrypt(data, Encoding.UTF8.GetBytes("metadata"))

					' Send the encrypted data
					Using stream As NetworkStream = New NetworkStream(socket)
						Using writer As StreamWriter = New StreamWriter(stream)
							writer.WriteLine(encryptedData)
							writer.Flush()
						End Using
					End Using
					ClipSyncDebug.log("Sent data: " + data)
					ClipSyncDebug.log("Sent encrypted data: " + encryptedData)
				Else
					ClipSyncDebug.log("[Communication] Clipboard null. Ignore")
				End If
			End Using
		Catch ex As Exception
			ClipSyncDebug.log("[Communication] Send clipboard data failed: " & ex.Message)
		End Try
	End Sub

	Public Shared Sub sendSyncRequest(ipAddr As IPAddress)
		Dim requestThread = New Thread(Sub()
										   sendRequestThread(ipAddr)
									   End Sub)
		requestThread.IsBackground = True
		requestThread.Start()
	End Sub

	Public Shared Sub sendClipboardData(ipAddr As IPAddress)
		Dim clipboardThread = New Thread(Sub()
											 sendClipboardThread(ipAddr)
										 End Sub)
		clipboardThread.IsBackground = True
		clipboardThread.Start()
	End Sub
End Class
